http://sourceware.org/ml/gdb-cvs/2010-08/msg00179.html

### src/gdb/ChangeLog	2010/08/27 21:03:39	1.12122
### src/gdb/ChangeLog	2010/08/27 23:09:58	1.12123
## -1,3 +1,8 @@
+2010-08-27  Doug Evans  <dje@google.com>
+
+	* dwarf2read.c (dw2_require_line_header): Read from .debug_types
+	as appropriate.
+
 2010-08-27  Tom Tromey  <tromey@redhat.com>
 
 	* dwarf2read.c (dwarf2_attach_fields_to_type): Don't set
--- src/gdb/dwarf2read.c	2010/08/27 21:03:39	1.447
+++ src/gdb/dwarf2read.c	2010/08/27 23:09:58	1.448
@@ -1990,6 +1990,7 @@
   struct attribute *attr;
   struct cleanup *cleanups;
   struct die_info *comp_unit_die;
+  struct dwarf2_section_info* sec;
   gdb_byte *beg_of_comp_unit, *info_ptr, *buffer;
   int has_children, i;
   struct dwarf2_cu cu;
@@ -2007,9 +2008,13 @@
 
   cleanups = make_cleanup (free_stack_comp_unit, &cu);
 
-  dwarf2_read_section (objfile, &dwarf2_per_objfile->info);
-  buffer_size = dwarf2_per_objfile->info.size;
-  buffer = dwarf2_per_objfile->info.buffer;
+  if (this_cu->from_debug_types)
+    sec = &dwarf2_per_objfile->types;
+  else
+    sec = &dwarf2_per_objfile->info;
+  dwarf2_read_section (objfile, sec);
+  buffer_size = sec->size;
+  buffer = sec->buffer;
   info_ptr = buffer + this_cu->offset;
   beg_of_comp_unit = info_ptr;
 
